diff -uNr qemu-3.1.0/linux-user/syscall.c qemu-3.1.0.mod/linux-user/syscall.c
--- qemu-3.1.0/linux-user/syscall.c	2018-12-11 19:44:34.000000000 +0200
+++ qemu-3.1.0.mod/linux-user/syscall.c	2018-12-24 21:26:44.589877092 +0200
@@ -45,7 +45,7 @@
 #include <sys/sem.h>
 #include <sys/statfs.h>
 #include <utime.h>
-#include <sys/sysinfo.h>
+#include <linux/sysinfo.h>
 #include <sys/signalfd.h>
 //#include <sys/user.h>
 #include <netinet/ip.h>
