diff -uNr termux-auth-1.1/termux-auth.h termux-auth-1.1.mod/termux-auth.h
--- termux-auth-1.1/termux-auth.h	2018-10-21 15:25:23.000000000 +0300
+++ termux-auth-1.1.mod/termux-auth.h	2018-12-21 15:41:14.783074938 +0200
@@ -3,11 +3,7 @@
 
 #include <stdbool.h>
 
-#ifdef __ANDROID__
 # define AUTH_HASH_FILE_PATH "/data/data/com.termux/files/home/.termux_authinfo"
-#else
-# define AUTH_HASH_FILE_PATH "/tmp/access_hash"
-#endif
 
 #ifdef  __cplusplus
 extern "C" {
