diff -uNr gnupg-2.2.11/tools/symcryptrun.c gnupg-2.2.11.mod/tools/symcryptrun.c
--- gnupg-2.2.11/tools/symcryptrun.c	2017-08-28 13:22:54.000000000 +0300
+++ gnupg-2.2.11.mod/tools/symcryptrun.c	2018-12-20 19:03:27.697256626 +0200
@@ -319,7 +319,7 @@
 
   p = getenv ("TMPDIR");
   if (!p || !*p)
-    p = "/tmp";
+    p = "@TERMUX_PREFIX@/tmp";
   if (p[strlen (p) - 1] == '/')
     name = xstrconcat (p, "gpg-XXXXXX", NULL);
   else
