diff -uNr dpkg-1.19.2/src/help.c dpkg-1.19.2.mod/src/help.c
--- dpkg-1.19.2/src/help.c	2018-10-08 01:07:52.000000000 +0300
+++ dpkg-1.19.2.mod/src/help.c	2018-12-17 18:36:20.595223965 +0200
@@ -127,7 +127,6 @@
     "update_dyld_shared_cache",
 #elif defined(__GLIBC__) || defined(__UCLIBC__) || \
       defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__NetBSD__)
-    "ldconfig",
 #endif
 #if BUILD_START_STOP_DAEMON
     "start-stop-daemon",
