diff -uNr gnupg-2.2.11/common/dotlock.c gnupg-2.2.11.mod/common/dotlock.c
--- gnupg-2.2.11/common/dotlock.c	2017-08-28 13:22:54.000000000 +0300
+++ gnupg-2.2.11.mod/common/dotlock.c	2018-12-20 19:18:51.697758376 +0200
@@ -605,6 +605,9 @@
 static int
 use_hardlinks_p (const char *tname)
 {
+
+  return 1;
+
   char *lname;
   struct stat sb;
   unsigned int nlink;
@@ -633,6 +636,7 @@
   unlink (lname);
   xfree (lname);
   return res;
+
 }
 #endif /*HAVE_POSIX_SYSTEM */
 
@@ -758,8 +762,6 @@
     }
   strcpy (stpcpy (h->lockname, file_to_lock), EXTSEP_S "lock");
   UNLOCK_all_lockfiles ();
-  if (h->use_o_excl)
-    my_debug_1 ("locking for '%s' done via O_EXCL\n", h->lockname);
 
   return h;
 
