diff -uNr gdb-8.0.1/readline/examples/rlfe/os.h gdb-8.0.1.mod/readline/examples/rlfe/os.h
--- gdb-8.0.1/readline/examples/rlfe/os.h	2017-06-04 18:51:27.000000000 +0300
+++ gdb-8.0.1.mod/readline/examples/rlfe/os.h	2018-12-21 19:45:23.151182368 +0200
@@ -288,7 +288,7 @@
 #   ifdef _PATH_UTMP
 #    define UTMPFILE	_PATH_UTMP
 #   else
-#    define UTMPFILE	"/etc/utmp"
+#    define UTMPFILE	"@TERMUX_PREFIX@/etc/utmp"
 #   endif /* _PATH_UTMP */
 #  endif
 # endif
