diff -uNr ttyrec-1.0.8/ttyrec.c ttyrec-1.0.8.mod/ttyrec.c
--- ttyrec-1.0.8/ttyrec.c	2006-06-11 18:52:50.000000000 +0300
+++ ttyrec-1.0.8.mod/ttyrec.c	2019-01-13 14:36:53.939372578 +0200
@@ -55,8 +55,9 @@
 #include <string.h>
 #include <stdlib.h>
 
-#if defined(SVR4)
 #include <fcntl.h>
+
+#if defined(SVR4)
 #include <stropts.h>
 #endif /* SVR4 */
 
@@ -156,7 +157,7 @@
 
 	shell = getenv("SHELL");
 	if (shell == NULL)
-		shell = "/bin/sh";
+		shell = "@TERMUX_PREFIX@/bin/sh";
 
 	getmaster();
 	fixtty();
@@ -203,11 +204,7 @@
 void
 finish()
 {
-#if defined(SVR4)
 	int status;
-#else /* !SVR4 */
-	union wait status;
-#endif /* !SVR4 */
 	register int pid;
 	register int die = 0;
 
