diff -uNr gawk-4.2.1/main.c gawk-4.2.1.mod/main.c
--- gawk-4.2.1/main.c	2018-02-23 10:44:11.000000000 +0200
+++ gawk-4.2.1.mod/main.c	2019-01-13 02:34:13.469351451 +0200
@@ -246,17 +246,6 @@
 	if ((cp = getenv("GAWK_LOCALE_DIR")) != NULL)
 		locale_dir = cp;
 
-#if defined(F_GETFL) && defined(O_APPEND)
-	// 1/2018: This is needed on modern BSD systems so that the
-	// inplace tests pass. I think it's a bug in those kernels
-	// but let's just work around it anyway.
-	int flags = fcntl(fileno(stderr), F_GETFL, NULL);
-	if (flags >= 0 && (flags & O_APPEND) == 0) {
-		flags |= O_APPEND;
-		(void) fcntl(fileno(stderr), F_SETFL, flags);
-	}
-#endif
-
 #if defined(LOCALEDEBUG)
 	initial_locale = locale;
 #endif
