diff -uNr apt-1.4.8/apt-pkg/contrib/fileutl.cc apt-1.4.8.mod/apt-pkg/contrib/fileutl.cc
--- apt-1.4.8/apt-pkg/contrib/fileutl.cc	2017-09-13 19:47:33.000000000 +0300
+++ apt-1.4.8.mod/apt-pkg/contrib/fileutl.cc	2018-12-17 23:10:27.311866580 +0200
@@ -106,7 +106,7 @@
             _exit(100);
       }
 
-      if (chdir("/tmp/") != 0)
+      if (chdir("@TERMUX_PREFIX@/tmp/") != 0)
 	 _exit(100);
 	 
       unsigned int Count = 1;
@@ -2790,10 +2790,10 @@
    struct stat st;
    if (!tmpdir || strlen(tmpdir) == 0 || // tmpdir is set
 	 stat(tmpdir, &st) != 0 || (st.st_mode & S_IFDIR) == 0) // exists and is directory
-      tmpdir = "/tmp";
+      tmpdir = "@TERMUX_PREFIX@/tmp";
    else if (geteuid() != 0 && // root can do everything anyway
 	 faccessat(AT_FDCWD, tmpdir, R_OK | W_OK | X_OK, AT_EACCESS) != 0) // current user has rwx access to directory
-      tmpdir = "/tmp";
+      tmpdir = "@TERMUX_PREFIX@/tmp";
 
    return string(tmpdir);
 }
@@ -3064,7 +3064,7 @@
       setenv("LOGNAME", pw->pw_name, 1);
       auto const shell = flNotDir(pw->pw_shell);
       if (shell == "false" || shell == "nologin")
-	 setenv("SHELL", "/bin/sh", 1);
+	 setenv("SHELL", "@TERMUX_PREFIX@/bin/sh", 1);
       else
 	 setenv("SHELL", pw->pw_shell, 1);
       auto const apt_setenv_tmp = [](char const * const env) {
