diff -uNr apt-1.4.8/apt-pkg/deb/dpkgpm.cc apt-1.4.8.mod/apt-pkg/deb/dpkgpm.cc
--- apt-1.4.8/apt-pkg/deb/dpkgpm.cc	2017-09-13 19:47:33.000000000 +0300
+++ apt-1.4.8.mod/apt-pkg/deb/dpkgpm.cc	2018-12-17 23:13:25.143722091 +0200
@@ -179,7 +179,7 @@
    so we use the binary from util-linux */
 static bool ionice(int PID)
 {
-   if (!FileExists("/usr/bin/ionice"))
+   if (!FileExists("@TERMUX_PREFIX@/bin/applets/ionice"))
       return false;
    pid_t Process = ExecFork();
    if (Process == 0)
@@ -187,7 +187,7 @@
       char buf[32];
       snprintf(buf, sizeof(buf), "-p%d", PID);
       const char *Args[4];
-      Args[0] = "/usr/bin/ionice";
+      Args[0] = "@TERMUX_PREFIX@/bin/applets/ionice";
       Args[1] = "-c3";
       Args[2] = buf;
       Args[3] = 0;
@@ -483,7 +483,7 @@
 
 	 debSystem::DpkgChrootDirectory();
 	 const char *Args[4];
-	 Args[0] = "/bin/sh";
+	 Args[0] = "@TERMUX_PREFIX@/bin/sh";
 	 Args[1] = "-c";
 	 Args[2] = Opts->Value.c_str();
 	 Args[3] = 0;
@@ -1281,8 +1281,6 @@
 	    sigemptyset(&d->sigmask);
 	    sigaddset(&d->sigmask, SIGTTOU);
 	    sigprocmask(SIG_BLOCK,&d->sigmask, &d->original_sigmask);
-	    if (tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw_tt) == -1)
-	       _error->Errno("tcsetattr", "Setting in Start via TCSAFLUSH for stdin failed!");
 	    sigprocmask(SIG_SETMASK, &d->original_sigmask, NULL);
 
 	 }
@@ -1361,8 +1359,6 @@
    }
    if(d->master >= 0)
    {
-      if (d->tt_is_valid == true && tcsetattr(STDIN_FILENO, TCSAFLUSH, &d->tt) == -1)
-	 _error->FatalE("tcsetattr", "Setting in Stop via TCSAFLUSH for stdin failed!");
       close(d->master);
       d->master = -1;
    }
@@ -2403,35 +2399,6 @@
       fprintf(report, " %s: %s\n", pkgname.c_str(), opstr);
    }
 
-   // attach dmesg log (to learn about segfaults)
-   if (FileExists("/bin/dmesg"))
-   {
-      fprintf(report, "Dmesg:\n");
-      FILE *log = popen("/bin/dmesg","r");
-      if(log != NULL)
-      {
-	 char buf[1024];
-	 while( fgets(buf, sizeof(buf), log) != NULL)
-	    fprintf(report, " %s", buf);
-	 pclose(log);
-      }
-   }
-
-   // attach df -l log (to learn about filesystem status)
-   if (FileExists("/bin/df"))
-   {
-
-      fprintf(report, "Df:\n");
-      FILE *log = popen("/bin/df -l","r");
-      if(log != NULL)
-      {
-	 char buf[1024];
-	 while( fgets(buf, sizeof(buf), log) != NULL)
-	    fprintf(report, " %s", buf);
-	 pclose(log);
-      }
-   }
-
    fclose(report);
 
 }
