diff -uNr gnupg-2.2.11/agent/agent.h gnupg-2.2.11.mod/agent/agent.h
--- gnupg-2.2.11/agent/agent.h	2018-06-12 14:25:39.000000000 +0300
+++ gnupg-2.2.11.mod/agent/agent.h	2018-12-20 19:00:21.073524080 +0200
@@ -339,7 +339,7 @@
   agent_Lunderscore (ctrl_t ctrl, const char *string)           \
   {                                                             \
     return ctrl? i18n_localegettext (ctrl->lc_messages, string) \
-      /*     */: gettext (string);                              \
+      /*     */: string;                              \
   }
 #else
 #define L_(a) (a)
