diff -uNr dropbear-2018.76/svr-authpubkey.c dropbear-2018.76.mod/svr-authpubkey.c
--- dropbear-2018.76/svr-authpubkey.c	2018-02-27 16:25:12.000000000 +0200
+++ dropbear-2018.76.mod/svr-authpubkey.c	2018-12-21 23:02:21.242666626 +0200
@@ -301,8 +301,6 @@
 	buffer * line = NULL;
 	unsigned int len;
 	int line_num;
-	uid_t origuid;
-	gid_t origgid;
 
 	TRACE(("enter checkpubkey"))
 
@@ -329,21 +327,8 @@
 	snprintf(filename, len + 22, "%s/.ssh/authorized_keys", 
 				ses.authstate.pw_dir);
 
-	/* open the file as the authenticating user. */
-	origuid = getuid();
-	origgid = getgid();
-	if ((setegid(ses.authstate.pw_gid)) < 0 ||
-		(seteuid(ses.authstate.pw_uid)) < 0) {
-		dropbear_exit("Failed to set euid");
-	}
-
 	authfile = fopen(filename, "r");
 
-	if ((seteuid(origuid)) < 0 ||
-		(setegid(origgid)) < 0) {
-		dropbear_exit("Failed to revert euid");
-	}
-
 	if (authfile == NULL) {
 		goto out;
 	}
